Add well-formedness test for assume/assert

xbauch opened the pull request:
Two simple tests are added: 1) targets are empty, 2) the guard evaluates to a
Boolean. The second test is a case-analysis of the expression id.

danpoe committed:
Add and use validation methods in equal_exprt

danpoe committed:
Add and use validation methods in code_assignt

danpoe committed:
Add and use validation methods in unsignedbv_typet

danpoe committed:
Add macros for goto model validation

danpoe committed:
Add templates to call check(), validate(), or validate_full() on a given object

danpoe committed:
Add functions to validate expressions

danpoe committed:
Add functions to validate types

danpoe committed:
Add functions to validate statements

danpoe committed:
Add validation methods to exprt

danpoe committed:
Add validation methods to typet

danpoe committed:
Add validation methods to codet

danpoe committed:
Add validate() methods to (components of) goto models

danpoe committed:
Validate goto models in cbmc (if enabled via command line flag)

danpoe committed:
Add check for well-formedness of assign instructions

danpoe committed:
Add new constructor to pass diagnostic information to incorrect_goto_program_exceptiont

danpoe committed:
Validate goto models in goto-analyzer (if enabled via command line flag)

danpoe committed:
Validate goto models in goto-instrument (if enabled via command line flags)

There are two flags: --validate-goto-binary reads in a goto binary and checks
that it is well formed and then exits; --validate-goto-model enables additional
well-formedness checks at various points during the execution of goto-instrument
in addition to the normal analyses that it runs.

Commit:
Add well-formedness test for assume/assert

Two simple tests are added: 1) targets are empty, 2) the guard evaluates to a
Boolean. The second test is a case-analysis of the expression id.

Commit:
Fix pre-pending a expr type check

Commit:
Added unit test

Checks that assert condition evaluates to bool.

danpoe commented on src/goto-programs/goto_program.cpp:
```diff
@@ -668,6 +668,91 @@ bool goto_programt::instructiont::equals(const instructiont &other) const
+    DATA_CHECK_WITH_DIAGNOSTICS(
+      evaluates_to_boolean(guard),
+      "assuming non-boolean condition\n" + guard.pretty(),
+      source_location);
```
Btw, if the instruction that we're checking has a corresponding code class in `std_code.h` that is used in the `code` field of the instruction, we should also add `check()/validate()/validate_full()` methods to that class.

For example, an assignment instruction contains a `code_assignt` object in its `code` field. However, many instructions don't actually fill in the `code` field I think. For example, I think an assume instruction does not contain a `code_assumet` object in its `code` field (as the `code_assumet` class is mostly only used to represent the result of parsing C code, but for the actual instruction it is not needed as it already has a `guard` which can hold the condition of the `assume`). In such cases we do not need to add the `check()/validate()/validate_full()` methods to `code_assumet`.

Commit:
Added checks for goto code

Check that gotos have targets, targets are targets, and goto guards evaluate to
a Boolean value. Also one unit test for the target validity.

Commit:
Added checks for function calls and returns

Function calls check type equality between function and returned value. Returns
only check the number of operands. Added one unit test for the function calls.

danpoe commented on src/goto-programs/goto_program.cpp:
```diff
@@ -755,13 +755,29 @@ void goto_programt::instructiont::validate(
+      code.get_statement() == ID_function_call,
+      "function call instruction should contain a call statement",
+      source_location);
+    code.validate(ns, vm);
```
The method `validate_full()` of `code` is already called by the call to `validate_code_full()` above so I think we can omit this one.

danpoe commented on src/goto-programs/goto_program.cpp:
```diff
@@ -755,13 +755,29 @@ void goto_programt::instructiont::validate(
+      code.get_statement() == ID_return,
+      "return instruction should contain a return statement",
+      source_location);
+    code.validate(ns, vm);
```
Can be omitted.

danpoe commented on src/util/std_code.h:
```diff
@@ -1052,6 +1052,40 @@ class code_function_callt:public codet
+  void check(const validation_modet vm = validation_modet::INVARIANT) const
+  {
+    DATA_CHECK(
+      operands().size() == 3, "function calls must have three operand");
```
`operand` -> `operands`

danpoe commented on src/util/std_code.h:
```diff
@@ -1052,6 +1052,40 @@ class code_function_callt:public codet
+    check(vm);
+    if(lhs().id() != ID_nil)
+      DATA_CHECK(
+        base_type_eq(lhs().type(), function().type(), ns),
```
I think we need to use `to_code_type(function().type()).return_type()` here.

Also we could additionally check that if `lhs.id() == ID_nil` then the return type of the call is `ID_empty` (which represents `void`).

xbauch replied:
I actually managed to trigger this type inconsistency in my unit test but I will try with the case as well.

danpoe commented on src/util/std_code.h:
```diff
@@ -1052,6 +1052,40 @@ class code_function_callt:public codet
+    for(const exprt &arg : op2().operands())
+    {
+      validate_expr_full_pick(arg, ns, vm);
+    }
```
I think we can omit this loop as the iteration over `op2().operands()` would already be done by the loop above which recursively calls `validate_full()` for all the operands of the expression.

danpoe commented on unit/goto-programs/goto_program_function_call.cpp:
```diff
@@ -0,0 +1,82 @@
+  GIVEN("A program with one function call")
+  {
+    symbol_tablet symbol_table;
+    const typet type1 = signedbv_typet(32);
```
I'd prefer using `const signedbv_typet type1` here. (though ultimately we'll have to assign this type to a `typet` anyways when assigining to the `type` field of a `symbolt`).

danpoe reviewed (commented):
Looks good to me overall. A few minor comments.

Commit:
Fix based on comments

Commit:
Added check for dead and decl

Plus a unit test. Had to pass symbol table down to goto_program.
